client/cli/src/runner.rs

```diff
@@ -177,45 +179,61 @@ impl<C: SubstrateCli> Runner<C> {
 
 	/// A helper function that runs a future with tokio and stops if the process receives the signal
 	/// `SIGTERM` or `SIGINT`.
-	pub fn run_subcommand<BU, B, BA, IQ, CL>(self, subcommand: &Subcommand, builder: BU)
-		-> Result<()>
+	pub fn run_subcommand<B, FBA, IQ, FCL, FB, LB, LCL, LBA>(
+		self, subcommand: &Subcommand, full_builder: FB, light_builder: LB,
+	) -> Result<()>
 	where
-		BU: FnOnce(Configuration)
-			-> sc_service::error::Result<(Arc<CL>, Arc<BA>, IQ, TaskManager)>,
+		FB: FnOnce(Configuration)
+			-> sc_service::error::Result<(Arc<FCL>, Arc<FBA>, IQ, TaskManager)>,
+		LB: FnOnce(Configuration)
+			-> sc_service::error::Result<(Arc<LCL>, Arc<LBA>, NetworkStatusSinks<B>, TaskManager)>,
 		B: BlockT + for<'de> serde::Deserialize<'de>,
-		BA: sc_client_api::backend::Backend<B> + 'static,
+		FBA: sc_client_api::backend::Backend<B> + 'static,
 		IQ: sc_service::ImportQueue<B> + 'static,
 		<B as BlockT>::Hash: FromStr,
 		<<B as BlockT>::Hash as FromStr>::Err: Debug,
 		<<<B as BlockT>::Header as HeaderT>::Number as FromStr>::Err: Debug,
-		CL: UsageProvider<B> + BlockBackend<B> + StorageProvider<B, BA> + Send + Sync +
+		FCL: UsageProvider<B> + BlockBackend<B> + StorageProvider<B, FBA> + Send + Sync +
 		'static,
+		LCL: HeaderBackend<B>,
+		LBA: sc_service::chain_ops::MaybeChtRootStorageProvider<B>,
 	{
 		let chain_spec = self.config.chain_spec.cloned_box();
 		let network_config = self.config.network.clone();
 		let db_config = self.config.database.clone();
 
 		match subcommand {
-			Subcommand::BuildSpec(cmd) => cmd.run(chain_spec, network_config),
+			Subcommand::BuildSpec(cmd) => {
+				if cmd.export_sync_state {
```

Comment:
No please. Please stop adding hacks over hacks.

Reply:
If we don't do this, we have to either pass `light_builder` to `cmd.run` or build the light client/backend/etc. even if it's not going to be used.

Reply:
Actually :thinking: building the light client backend even if it's not used might not be a bad idea. Of course it's not optimal but that just slow down the initialization. If it can improve the situation in term of readability/maintenance, it might be a good trade

Reply:
As I said below, just throw an error when the user started the node as a full node. This is still better than this.